src/components/Login.js

```diff
@@ -0,0 +1,222 @@
+  messagingSenderId: process.env.REACT_APP_FIREBASE_MESSAGINGSENDERID,
+  appId: process.env.REACT_APP_FIREBASE_APPID,
+  measurementId: process.env.REACT_APP_FIREBASE_MEASUREMENTID,
+};
+
+// Initialize Firebase
+const app = initializeApp(firebaseConfig);
+const auth = getAuth();
+const db = getFirestore(app);
+
+function Login(props) {
+  const SIGN_IN = "sign_in";
+  const SIGN_UP = "sign_up";
+
+  const [isShowSignWhat, setIsShowSignWhat] = useState();
+  const [userEmail, setUserEmail] = useState("");
+  const [userPassword, setUserPassword] = useState("");
+  const [userName, setUserName] = useState("");
+  const [userPhoto, setUserPhoto] = useState("");
+  const [userRole, setUserRole] = useState("");
+  const [userLink, setUserLink] = useState("");
+
+  const redirect = useNavigate();
+
+  const showSingIn = () => {
+    setIsShowSignWhat(SIGN_IN);
+  };
+
+  const showSingUp = () => {
+    setIsShowSignWhat(SIGN_UP);
+  };
+
+  const signUp = () => {
+    createUserWithEmailAndPassword(auth, userEmail, userPassword)
+      .then((userCredential) => {
+        const user = userCredential.user;
+        writeUserData(user.uid);
+        getUserData(user.uid);
+      })
+      .catch((error) => {
+        const errorCode = error.code;
+        const errorMessage = error.message;
+        console.log("errorCode", errorCode);
+        console.log("errorMessage", errorMessage);
+
+        switch (errorCode) {
+          case "auth/wrong-password":
+            alert("密碼錯誤喔");
+            break;
+          case "auth/user-not-found":
+            alert("帳號不存在喔");
+            break;
+          case "auth/weak-password":
+            alert("密碼需至少六碼喔");
+            break;
+          default:
+            alert("登入失敗QQ");
+        }
+      });
+  };
+
+  const signIn = () => {
+    signInWithEmailAndPassword(auth, userEmail, userPassword)
+      .then((userCredential) => {
+        const user = userCredential.user;
+        getUserData(user.uid);
+      })
+      .catch((error) => {
+        const errorCode = error.code;
+        const errorMessage = error.message;
+        console.error("errorCode", errorCode);
+        console.error("errorMessage", errorMessage);
+
+        switch (errorCode) {
+          case "auth/wrong-password":
+            alert("密碼錯誤喔");
+            break;
+          case "auth/user-not-found":
+            alert("帳號不存在喔");
+            break;
+          default:
+            alert("登入失敗QQ");
+        }
+      });
+  };
+
+  const writeUserData = (userId) => {
+    setDoc(doc(db, "user/" + userId), {
+      email: userEmail,
+      password: userPassword,
+      name: userName,
+      link: userLink,
+      uid: userId,
+      follower: [],
+      following: [],
+      pic: userPhoto,
+      role: userRole,
+    });
+    console.log("data written to database");
+  };
+
+  const getUserData = (userId) => {
+    const unsub = onSnapshot(doc(db, "user/" + userId), (doc) => {
```

Comment:
1. What is 'unsub'?
2. ${}

Reply:
1. rename the firebase onSnapshot
2. modified